Retain test mode context in CRON jobs queued up while checking out

lovo-h opened the pull request:
* Fixes N/A

#### Changes proposed in this Pull Request
* Before, the test mode context would be lost when the CRON job was dequeued. Now, the live mode context is passed along and applied in the CRON job.

<!--
Title: A descriptive, yet concise, title.
-->

<!--
Description: Write a brief summary about this PR. As you compose your summary, consider each of these questions and address them if appropriate. Why is this change needed? What does this change do? Were there other solutions you considered? Why did you choose to pursue this solution? Describe any trade-offs you might have had to make.
-->

<!--
Questions for PR author:
- How can this code break?
- What are we doing to make sure this code doesn't break?
-->

<!--
Images or gifs: Include before and after screenshots or gifs/videos when it makes sense.
-->

#### Testing instructions

<!--
Testing instructions: How should this be tested and how can a reviewer test the end-user functionality? Are there known issues that you plan to address in a future PR? Are there any side effects that readers should be aware of?
-->

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

**Checkout through WooPay**
* Ensure adding fee details to order notes async applies correctly in test/live mode (#4429).
* Ensure updating saved payment method async applies correctly in test/live mode (#4443).
* Ensure updating existing customer details applies correctly in test/live mode (#4495).

**Checkout through the merchant site using a saved payment method**
* Ensure adding fee details to order notes async applies correctly in test/live mode (#4429).
* Ensure updating saved payment method async applies correctly in test/live mode (#4443).
* Ensure updating existing customer details applies correctly in test/live mode (#4495).

-------------------

- [x] Run `npm run changelog` to add a changelog file, choose `patch` to leave it empty if the change is not significant. You can add multiple changelog files in one PR by running this command a few times. 
- [x] Covered with tests (or have a good reason not to test in description ☝️)
- [x] Tested on mobile (or does not apply)

**Post merge**

<!--
Make sure you edit the page for the current release when adding testing instructions.
We often create a blank page ahead of time for the next release.
If this PR need not be QA tested, edit to 'QA Testing Not Applicable'
-->

- [ ] Link to testing instructions from [release testing doc](https://github.com/Automattic/woocommerce-payments/wiki/Release-testing-instructions) following [these instructions](https://github.com/Automattic/woocommerce-payments/wiki/How-to-write-good-manual-testing-scenarios) : _Add link here / 'QA Testing Not Applicable'_
- [ ] Add or update [critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Critical-flows) and [testing instructions for critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Testing-instructions-for-critical-flows), if applicable.
- [ ] Add what's changed (description, screenshot, demo videos etc.) to the release announcement post, if applicable.

lovo-h committed:
Retain test mode context in 'update_saved_payment_method()' CRON job

lovo-h committed:
Retain test mode context in 'add_fee_breakdown_to_order_notes()' CRON job

lovo-h committed:
Retain test mode context in 'update_customer_with_order_data()' CRON job

lovo-h committed:
Ensure the correct customer email's applied when updating in CRON job

Before, 'wp_get_current_user()' returned user with ID of 0. So, when
$user->user_email was called, 'false' would be returned. Now, the user
from the $order is used, instead.

lovo-h committed:
Fix test: 'test_updates_payment_method_billing_details'

Update the expected arguments for 'wcpay_update_saved_payment_method'.

lovo-h committed:
Add changelog: Retain test mode context in CRON jobs enqueued while checking out.

deepakpathania commented on includes/class-wc-payment-gateway-wcpay.php:
```diff
@@ -1103,18 +1103,39 @@ protected function prepare_payment_information( $order ) {
-	public function update_customer_with_order_data( $order_id, $customer_id ) {
+	 * @param array  $options        Additional options to apply.
+	 */
+	public function update_customer_with_order_data( $order_id, $customer_id, $options = [] ) {
```
Nitpick but it might be better to pass the exact params to the async functions instead of an options array here for clear reference in the function signature.

lovo-h replied:
Since there are only 2 additional params, this makes sense. Fixed in [9785250](https://github.com/Automattic/woocommerce-payments/pull/4713/commits/9785250eb8f8ebd642102582195d21f78c4619f7).

deepakpathania replied:
Don't we have to update the accepted params in the `add_action` call for this method for the extra param to be reflected? 🤔  @lovo-h

lovo-h replied:
Good catch. Fixed in [f8365a0](https://github.com/Automattic/woocommerce-payments/pull/4713/commits/f8365a0085d6c97ef3f78a1c608605d5a57ceb4c). Also, added a test in [6facffb](https://github.com/Automattic/woocommerce-payments/pull/4713/commits/6facffb68171625ca3aa896d1081ff0eaf69e901), to avoid a similar issue in the future.

deepakpathania reviewed (commented):
I tested only the merchant checkout flow and that seems to be working fine, left a small nitpick comment.

lovo-h committed:
Merge branch 'develop' into fix/1257-test-data-leaking-to-live

lovo-h committed:
Replace $options with exact params for clarity

lovo-h committed:
Fix the number of accepted arguments from 3 to 4

lovo-h committed:
Replace literal action name with constant 'UPDATE_SAVED_PAYMENT_METHOD'

lovo-h committed:
Add test to ensure UPDATE_CUSTOMER_WITH_ORDER_DATA job scheduled correctly

EvanDotPro reviewed (approved):
Tested with WooPay and seems to be working :+1:.

reykjalin commented on includes/class-wc-payment-gateway-wcpay.php:
```diff
@@ -1194,7 +1228,10 @@ public function process_payment_for_order( $cart, $payment_information, $additio
 
-		list( $user, $customer_id ) = $this->manage_customer_details_for_order( $order );
+		$customer_details_options   = [
+			'is_woopay' => $metadata['paid_on_woopay'] ?? false,
```
I think the metadata may be serialized into a `'1'` in some cases so it's probably worth it to `filter_var( $metadata['paid_on_woopay'], FILTER_VALIDATE_BOOLEAN ) ?? false;` here 🤔

lovo-h replied:
Good call. Fixed in [1a88ed4](https://github.com/Automattic/woocommerce-payments/pull/4713/commits/1a88ed48cc8e663c331bf475c79cb5b72b3b8d9d).

reykjalin reviewed (changes requested):
I had one minor comment but otherwise this is looking good 🎉

deepakpathania reviewed (approved):
Thanks for addressing the comment, changes LGTM now. (tested only merchant site)  ✅

lovo-h committed:
Ensure 'is_woopay' resolves to bool

Before, it was possible that $metadata['paid_on_woopay'] could be
deserialized to a string type. Now, we ensure it resolves to a bool type.

lovo-h committed:
Fix test_updates_payment_method_billing_details: expect anything for time()

reykjalin reviewed (approved):
LGTM 🚀
